encoding/ccf/decode.go

```diff
@@ -0,0 +1,1887 @@
+
+	case cadence.MetaType:
+		return d.decodeTypeValue()
+
+	case *cadence.CapabilityType:
+		return d.decodeCapability(typ)
+
+	case *cadence.EnumType:
+		return d.decodeEnum(typ, types)
+
+	default:
+		err := decodeCBORTagWithKnownNumber(d.dec, CBORTagTypeAndValue)
+		if err != nil {
+			return nil, fmt.Errorf("unexpected encoded value of Cadence type %s (%T): %s", typ.ID(), typ, err.Error())
+		}
+
+		// Decode ccf-type-and-value-message.
+		return d.decodeTypeAndValue(types)
+	}
+}
+
+// decodeVoid decodes encoded void-value as
+// language=CDDL
+// void-value = nil
+func (d *Decoder) decodeVoid() (cadence.Value, error) {
+	err := d.dec.DecodeNil()
+	if err != nil {
+		return nil, err
+	}
+	return cadence.NewMeteredVoid(d.gauge), nil
+}
+
+// decodeBool decodes encoded bool-value as
+// language=CDDL
+// bool-value = bool
+func (d *Decoder) decodeBool() (cadence.Value, error) {
+	b, err := d.dec.DecodeBool()
+	if err != nil {
+		return nil, err
+	}
+	return cadence.NewMeteredBool(d.gauge, b), nil
+}
+
+// decodeCharacter decodes encoded character-value as
+// language=CDDL
+// character-value = tstr
+func (d *Decoder) decodeCharacter() (cadence.Value, error) {
+	s, err := d.dec.DecodeString()
+	if err != nil {
+		return nil, err
+	}
+	return cadence.NewMeteredCharacter(
+		d.gauge,
+		common.NewCadenceCharacterMemoryUsage(len(s)),
+		func() string {
+			return s
+		})
+}
+
+// decodeString decodes encoded string-value as
+// language=CDDL
+// string-value = tstr
+// NOTE: invalid UTF-8 is rejected.
+func (d *Decoder) decodeString() (cadence.Value, error) {
+	s, err := d.dec.DecodeString()
+	if err != nil {
+		return nil, err
+	}
+
+	return cadence.NewMeteredString(
+		d.gauge,
+		common.NewCadenceStringMemoryUsage(len(s)),
+		func() string {
+			return s
+		},
+	)
+}
+
+// decodeAddress decodes address-value as
+// language=CDDL
+// address-value = bstr .size 8
+func (d *Decoder) decodeAddress() (cadence.Value, error) {
+	b, err := d.dec.DecodeBytes()
+	if err != nil {
+		return nil, err
+	}
+	if len(b) != 8 {
+		return nil, fmt.Errorf("encoded address-value has length %d (expected 8 bytes)", len(b))
+	}
+	return cadence.BytesToMeteredAddress(d.gauge, b), nil
+}
+
+// decodeInt decodes int-value as
+// language=CDDL
+// int-value = bigint
+func (d *Decoder) decodeInt() (cadence.Value, error) {
+	bigInt, err := d.dec.DecodeBigInt()
+	if err != nil {
+		return nil, err
+	}
+
+	return cadence.NewMeteredIntFromBig(
+		d.gauge,
+		common.NewCadenceIntMemoryUsage(
+			common.BigIntByteLength(bigInt),
+		),
+		func() *big.Int {
+			return bigInt
+		},
+	), nil
+}
+
+// decodeInt8 decodes int8-value as
+// language=CDDL
+// int8-value = (int .ge -128) .le 127
+func (d *Decoder) decodeInt8() (cadence.Value, error) {
+	i, err := d.dec.DecodeInt64()
+	if err != nil {
+		return nil, err
+	}
+	if i < math.MinInt8 || i > math.MaxInt8 {
+		return nil, fmt.Errorf(
+			"encoded int8-value %d is outside range of Int8 [%d, %d]",
+			i,
+			math.MaxInt8,
```

Comment:
```suggestion
			math.MinInt8,
```

Reply:
Good catch!  Done in bc220dce48e3811da284fd94fd47e5be21c810de.